cosi/cosi.go

```diff
@@ -1,450 +1,399 @@
-			if cm.mask[byt]&bit == 0 {
-				cm.mask[byt] |= bit // disable it
-				cm.aggPublic.Sub(cm.aggPublic, cm.publics[i])
-			}
-		} else {
-			// Participant i enabled in new mask.
-			if cm.mask[byt]&bit != 0 {
-				cm.mask[byt] &^= bit // enable it
-				cm.aggPublic.Add(cm.aggPublic, cm.publics[i])
-			}
-		}
+	m := make([]byte, len(a))
+	for i := range m {
+		m[i] = a[i] | b[i]
 	}
-	return nil
+	return m, nil
 }
 
-// MaskLen returns the length in bytes
-// of a complete disable-mask for this cosigner list.
-func (cm *mask) MaskLen() int {
-	return (len(cm.publics) + 7) >> 3
+// Policy represents a fully customizable cosigning policy deciding what
+// cosigner sets are and aren't sufficient for a collective signature to be
+// considered acceptable to a verifier. The Check method may inspect the set of
+// participants that cosigned by invoking cosi.Mask and/or cosi.MaskBit, and may
+// use any other relevant contextual information (e.g., how security-critical
+// the operation relying on the collective signature is) in determining whether
+// the collective signature was produced by an acceptable set of cosigners.
+type Policy interface {
+	Check(m *Mask) bool
 }
 
-// SetMaskBit enables or disables the mask bit for an individual cosigner.
-func (cm *mask) SetMaskBit(signer int, enabled bool) {
-	if signer > len(cm.publics) {
-		panic("SetMaskBit range out of index")
-	}
-	byt := signer >> 3
-	bit := byte(1) << uint(signer&7)
-	if !enabled {
-		if cm.mask[byt]&bit == 0 { // was enabled
-			cm.mask[byt] |= bit // disable it
-			cm.aggPublic.Sub(cm.aggPublic, cm.publics[signer])
-		}
-	} else { // enable
-		if cm.mask[byt]&bit != 0 { // was disabled
-			cm.mask[byt] &^= bit
-			cm.aggPublic.Add(cm.aggPublic, cm.publics[signer])
-		}
-	}
+// CompletePolicy is the default policy requiring that all participants have
+// cosigned to make a collective signature valid.
+type CompletePolicy struct {
 }
 
-// MaskBit returns a boolean value indicating whether
-// the indicated signer is enabled (true) or disabled (false)
-func (cm *mask) MaskBit(signer int) bool {
-	if signer > len(cm.publics) {
-		panic("MaskBit given index out of range")
-	}
-	byt := signer >> 3
-	bit := byte(1) << uint(signer&7)
-	return (cm.mask[byt] & bit) != 0
+// Check verifies that all participants have contributed to a collective
+// signature.
+func (p CompletePolicy) Check(m *Mask) bool {
+	return m.CountEnabled() == m.CountTotal()
 }
 
-// bytes returns the byte representation of the mask
-// The bits that are left are set to a default value (1) for
-// non malleability.
-func (cm *mask) bytes() []byte {
-	clone := make([]byte, len(cm.mask))
-	for i := range clone {
-		clone[i] = 0xff
-	}
-	copy(clone[:], cm.mask)
-	return clone
+// ThresholdPolicy allows to specify a simple t-of-n policy requring that at
+// least the given threshold number of participants have cosigned to make a
+// collective signature valid.
+type ThresholdPolicy struct {
+	t int
```

Comment:
Can you instantiate a `ThresholdPolicy` from outside of `package cosi` with `t` being private? I don't think so.